Lib/compression/zstd/__init__.py

```diff
@@ -0,0 +1,293 @@
+                raise  # Error on the first iteration; bail out.
+        results.append(res)
+        if not decomp.eof:
+            raise ZstdError("Compressed data ended before the "
+                            "end-of-stream marker was reached")
+        data = decomp.unused_data
+        if not data:
+            break
+    return b"".join(results)
+
+class _UnsupportedCParameter:
+    def __set_name__(self, _, name):
+        self.name = name
+
+    def __get__(self, *_, **__):
+        msg = ("%s CParameter not available, zstd version is %s.") % (
+            self.name,
+            zstd_version,
+        )
+        raise NotImplementedError(msg)
+
+
+class CParameter(enum.IntEnum):
+    """Compression parameters"""
+
+    compressionLevel = _zstd._ZSTD_c_compressionLevel
+    windowLog = _zstd._ZSTD_c_windowLog
+    hashLog = _zstd._ZSTD_c_hashLog
+    chainLog = _zstd._ZSTD_c_chainLog
+    searchLog = _zstd._ZSTD_c_searchLog
+    minMatch = _zstd._ZSTD_c_minMatch
+    targetLength = _zstd._ZSTD_c_targetLength
+    strategy = _zstd._ZSTD_c_strategy
+
+    targetCBlockSize = _UnsupportedCParameter()
+
+    enableLongDistanceMatching = _zstd._ZSTD_c_enableLongDistanceMatching
+    ldmHashLog = _zstd._ZSTD_c_ldmHashLog
+    ldmMinMatch = _zstd._ZSTD_c_ldmMinMatch
+    ldmBucketSizeLog = _zstd._ZSTD_c_ldmBucketSizeLog
+    ldmHashRateLog = _zstd._ZSTD_c_ldmHashRateLog
+
+    contentSizeFlag = _zstd._ZSTD_c_contentSizeFlag
+    checksumFlag = _zstd._ZSTD_c_checksumFlag
+    dictIDFlag = _zstd._ZSTD_c_dictIDFlag
+
+    nbWorkers = _zstd._ZSTD_c_nbWorkers
+    jobSize = _zstd._ZSTD_c_jobSize
+    overlapLog = _zstd._ZSTD_c_overlapLog
+
+    @functools.lru_cache(maxsize=None)
+    def bounds(self):
+        """Return lower and upper bounds of a compression parameter, both inclusive."""
+        # 1 means compression parameter
+        return _zstd._get_param_bounds(1, self.value)
+
+
+class DParameter(enum.IntEnum):
+    """Decompression parameters"""
+
+    windowLogMax = _zstd._ZSTD_d_windowLogMax
+
+    @functools.lru_cache(maxsize=None)
+    def bounds(self):
+        """Return lower and upper bounds of a decompression parameter, both inclusive."""
+        # 0 means decompression parameter
+        return _zstd._get_param_bounds(0, self.value)
+
+
+class Strategy(enum.IntEnum):
+    """Compression strategies, listed from fastest to strongest.
+
+    Note : new strategies _might_ be added in the future, only the order
+    (from fast to strong) is guaranteed.
+    """
+
+    fast = _zstd._ZSTD_fast
+    dfast = _zstd._ZSTD_dfast
+    greedy = _zstd._ZSTD_greedy
+    lazy = _zstd._ZSTD_lazy
+    lazy2 = _zstd._ZSTD_lazy2
+    btlazy2 = _zstd._ZSTD_btlazy2
+    btopt = _zstd._ZSTD_btopt
+    btultra = _zstd._ZSTD_btultra
+    btultra2 = _zstd._ZSTD_btultra2
+
+
+# Set CParameter/DParameter types for validity check
+_zstd._set_parameter_types(CParameter, DParameter)
+
+zstd_support_multithread = CParameter.nbWorkers.bounds() != (0, 0)
```

Comment:
```suggestion
zstd_support_multithreaded = CParameter.nbWorkers.bounds() != (0, 0)
```

Reply:
is there a use for this attribute at all?  the norm for a CPython build is going to be that libzstd always supports multithreaded compression/decompression internally so I don't expect users will write code checking if it can or not?

Reply:
Yes I agree it is rather uncommon to have zstd without multithreading. I suppose it might be the case on emscripten? But that's niche enough I think deleting this entirely is reasonable.

Reply:
Deleted, moved the detection to tests where it is used in a few places. We should probably decide if we want to support building without multithreading period.